### PR TITLE
[SPARK-24737][SQL] Type coercion between StructTypes.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

We can support type coercion between `StructType`s where all the internal types are compatible.

## How was this patch tested?

Added tests.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @gatorsmile @cloud-fan

**Comment by ueshin** [comment]
Jenkins, retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -185,6 +185,15 @@ object TypeCoercion {
               MapType(kt, vt, valueContainsNull1 || valueContainsNull2)
             }
           }
+        case (StructType(fields1), StructType(fields2)) if fields1.length == fields2.length =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: is it time to reduce code duplication now?
```
def mergeComplexTypes(
    t1: DataType,
    t2: DataType,
    mergeFunc: (DataType, DataType) => Option[DataType]): Option[DataType] = (t1, t2) match {
  case (ArrayType(et1, containsNull1), ArrayType(et2, containsNull2)) =>
    mergeFunc(et1, et2).map(ArrayType(_, containsNull1 || containsNull2))
  ...
}

```

**Comment by cloud-fan** [comment]
LGTM

**Comment by ueshin** [comment]
Jenkins, retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 2]
```diff
@@ -166,6 +166,30 @@ object TypeCoercion {
     case (l, r) => None
   }
 
+  private def mergeComplexTypes(
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: nit: I would use similar naming conversion such as `findWiderTypeForComplex`

**Review by HyukjinKwon — APPROVED** [review]
LGTM too

**Comment by viirya** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 3]
```diff
@@ -166,6 +166,30 @@ object TypeCoercion {
...
+    case (MapType(kt1, vt1, valueContainsNull1), MapType(kt2, vt2, valueContainsNull2)) =>
+      mergeFunc(kt1, kt2).flatMap { kt =>
+        mergeFunc(vt1, vt2).map { vt =>
+          MapType(kt, vt, valueContainsNull1 || valueContainsNull2)
+        }
+      }
+    case (StructType(fields1), StructType(fields2)) if fields1.length == fields2.length =>
+      val resolver = SQLConf.get.resolver
+      fields1.zip(fields2).foldLeft(Option(new StructType())) {
+        case (Some(struct), (field1, field2)) if resolver(field1.name, field2.name) =>
```
[the remark this answers is not in the document]
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Ah, I see. `findTightestCommonType` only picks first field name when caseSensitive = false. This does the same thing actually.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 4]
```diff
@@ -166,6 +166,30 @@ object TypeCoercion {
     case (l, r) => None
   }
 
+  private def mergeComplexTypes(
+      t1: DataType,
+      t2: DataType,
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Maybe `findTightestCommonType` can also reuse `mergeComplexTypes` ?

Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Seems like the behavior will be changed if we reuse there, and the test [TypeCoercionSuite.scala#L397-L400](https://github.com/apache/spark/blob/bed6849dbf51e1981772cd353ce1a7ae4f0626e2/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala#L397-L400) fails.

Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: The test result might be wrong?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: yea it's wrong after this patch

**Review by viirya — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master

**Pull request closed without merging**